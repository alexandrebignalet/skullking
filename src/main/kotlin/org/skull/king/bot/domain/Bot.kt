package org.skull.king.bot.domain

import com.fasterxml.jackson.databind.ObjectMapper
import org.skull.king.application.infrastructure.framework.command.Command
import org.skull.king.core.domain.state.AnnounceState
import org.skull.king.core.domain.state.RoundState
import org.skull.king.core.domain.state.SkullKingEvent
import org.skull.king.core.domain.state.Skullking
import org.skull.king.core.usecases.AnnounceWinningCardsFoldCountSaga
import org.skull.king.core.usecases.PlayCardSaga
import org.skull.king.game_room.domain.GameRoom
import org.skull.king.game_room.domain.GameUser.BotGameUser

data class Bot(
    val trigger: SkullKingEvent,
    val gameUser: BotGameUser,
    val state: State = Void()
) {

    companion object {
        fun from(room: GameRoom, event: SkullKingEvent) = room.bots.map { Bot(event, it) }
    }

    fun withContext(context: Skullking) = when (context) {
        is AnnounceState -> Bot(trigger, gameUser, Announce(context))
        is RoundState -> Bot(trigger, gameUser, PlayCard(context))
        else -> Bot(trigger, gameUser, Void(context))
    }

    fun command() = state.command(trigger, gameUser)
}

sealed interface State {
    fun command(trigger: SkullKingEvent, gameUser: BotGameUser): BotActionResult<Command<*>>
}

data class Announce(val game: AnnounceState) : State {

    override fun command(
        trigger: SkullKingEvent,
        gameUser: BotGameUser
    ): BotActionResult<AnnounceWinningCardsFoldCountSaga> {
        if (game.hasAlreadyAnnounced(gameUser.id)) {
            return BotActionResult.failure("${gameUser.name} has already played on ${trigger.javaClass.simpleName} for ${game::class.simpleName}:${game.version}")
        }

        val currentPlayer = game.players.find { it.id == gameUser.id }

        return BotActionResult.success(
            AnnounceWinningCardsFoldCountSaga(
                gameId = game.gameId,
                playerId = gameUser.id,
                count = if (currentPlayer == null) 1
                else gameUser.strategy.computeAnnounce(game, currentPlayer)
            )
        )
    }
}

data class PlayCard(val game: RoundState) : State {
    override fun command(trigger: SkullKingEvent, gameUser: BotGameUser): BotActionResult<PlayCardSaga> {
        val currentPlayer = game.currentPlayer()
        if (currentPlayer?.id != gameUser.id) {
            val objectMapper = ObjectMapper()
            val debugMessage =
                if (currentPlayer != null) "" else " : current is null -> ${objectMapper.writeValueAsString(game)}"
            return BotActionResult.failure("it is not ${gameUser.name} turn to play on ${trigger.javaClass.simpleName} for ${game.javaClass.simpleName}:${game.version} $debugMessage")
        }

        return BotActionResult.success(
            PlayCardSaga(
                gameId = game.gameId,
                playerId = gameUser.id,
                card = gameUser.strategy.computeCardPlay(game, currentPlayer)
            )
        )
    }
}

data class Void(val game: Skullking? = null) : State {
    override fun command(trigger: SkullKingEvent, gameUser: BotGameUser) =
        BotActionResult.failure<Command<*>>("${gameUser.name} has nothing to do on ${trigger::class.simpleName} for ${game?.javaClass?.simpleName}:${game?.version}")
}
