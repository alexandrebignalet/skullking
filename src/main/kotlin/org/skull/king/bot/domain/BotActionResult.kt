package org.skull.king.bot.domain

sealed interface BotActionResult<out T> {

    companion object {
        fun <T> success(result: T) = Success(result)
        fun <T> failure(failure: String) = Failure<T>(failure)
    }

    fun onFailure(block: (failure: String) -> Unit): BotActionResult<T> {
        if (this is Failure) {
            block(this.failure)
        }
        return this
    }

    fun onSuccess(block: (success: T) -> Unit): BotActionResult<T> {
        if (this is Success) {
            block(this.result)
        }
        return this
    }

    data class Success<T>(val result: T) : BotActionResult<T>
    data class Failure<T>(val failure: String) : BotActionResult<T>

}