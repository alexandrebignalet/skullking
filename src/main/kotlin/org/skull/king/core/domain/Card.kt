package org.skull.king.core.domain

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import java.util.*

enum class CardType {
    ESCAPE,
    MERMAID,
    COLORED,
    PIRATE,
    SCARY_MARY,
    SKULLKING,
    KRAKEN,
    WHITE_WHALE,
    BUTIN
}

@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.EXISTING_PROPERTY,
    property = "type",
    visible = true
)
@JsonSubTypes(
    JsonSubTypes.Type(value = ColoredCard::class, name = "COLORED"),
    JsonSubTypes.Type(value = SkullkingCard::class, name = "SKULLKING"),
    JsonSubTypes.Type(value = ScaryMary::class, name = "SCARY_MARY"),
    JsonSubTypes.Type(value = Pirate::class, name = "PIRATE"),
    JsonSubTypes.Type(value = Escape::class, name = "ESCAPE"),
    JsonSubTypes.Type(value = Mermaid::class, name = "MERMAID"),
    JsonSubTypes.Type(value = Butin::class, name = "BUTIN"),
    JsonSubTypes.Type(value = Kraken::class, name = "KRAKEN"),
    JsonSubTypes.Type(value = WhiteWhale::class, name = "WHITE_WHALE"),
)
sealed class Card(val type: CardType) : Comparable<Card> {
    abstract override operator fun compareTo(other: Card): Int

    open val id get(): String = type.name

    companion object {
        fun fromId(cardId: String, usage: ScaryMaryUsage): Card {
            val split = cardId.split("__")
            val type = CardType.valueOf(split.first())
            return when (type) {
                CardType.ESCAPE -> Escape
                CardType.MERMAID -> Mermaid(MermaidName.valueOf(split[1]))
                CardType.COLORED -> ColoredCard(split[2].toInt(), CardColor.valueOf(split[1]))
                CardType.PIRATE -> Pirate(PirateName.valueOf(split[1]))
                CardType.SCARY_MARY -> ScaryMary(usage)
                CardType.SKULLKING -> SkullkingCard
                CardType.KRAKEN -> Kraken
                CardType.WHITE_WHALE -> WhiteWhale
                CardType.BUTIN -> Butin
            }
        }
    }
}

data class ColoredCard(val value: Int, val color: CardColor) : Card(CardType.COLORED) {
    override fun compareTo(other: Card): Int {
        return when (other) {
            Butin -> 1
            is ColoredCard -> when (other.color) {
                color -> value.compareTo(other.value)
                CardColor.BLACK -> -1
                else -> 1
            }

            Escape -> 1
            Kraken -> -1
            is Mermaid -> -1
            is Pirate -> -1
            is ScaryMary -> if (other.usage == ScaryMaryUsage.ESCAPE) 1 else -1
            SkullkingCard -> -1
            WhiteWhale -> -1
        }
    }

    override val id get(): String = "${type}__${color}__$value"
}

enum class CardColor { RED, BLUE, YELLOW, BLACK, GREEN, PURPLE }

enum class PirateName {
    HARRY_THE_GIANT,
    TORTUGA_JACK,
    EVIL_EMMY,
    BADEYE_JOE,
    BETTY_BRAVE,

    ROSIE_LA_DOUCE,
    WILL_LE_BANDIT,
    RASCAL_LE_FLAMBEUR,
    JUANITA_JADE,
    HARRY_LE_GEANT
}

enum class MermaidName {
    NONE,
    CIRCE,
    ALYRA
}

data class Pirate(val name: PirateName) : Card(CardType.PIRATE) {
    override fun compareTo(other: Card): Int {
        return when (other) {
            Butin -> 1
            is ColoredCard -> 1
            Escape -> 1
            Kraken -> -1
            is Mermaid -> 1
            is Pirate -> 1
            is ScaryMary -> 1
            SkullkingCard -> -1
            WhiteWhale -> -1
        }
    }

    override val id get(): String = "${type}__$name"
}

object SkullkingCard : Card(CardType.SKULLKING) {
    override fun compareTo(other: Card): Int {
        return when (other) {
            Butin -> 1
            is ColoredCard -> 1
            Escape -> 1
            Kraken -> -1
            is Mermaid -> -1
            is Pirate -> 1
            is ScaryMary -> 1
            SkullkingCard -> 1
            WhiteWhale -> -1
        }
    }
}

data class Mermaid(val name: MermaidName = MermaidName.NONE) : Card(CardType.MERMAID) {
    override fun compareTo(other: Card): Int {
        return when (other) {
            Butin -> 1
            is ColoredCard -> 1
            Escape -> 1
            Kraken -> -1
            is Mermaid -> 1
            is Pirate -> -1
            is ScaryMary -> if (other.usage == ScaryMaryUsage.ESCAPE) 1 else -1
            SkullkingCard -> 1
            WhiteWhale -> -1
        }
    }

    override val id get(): String = "${type}__$name"
}

object Escape : Card(CardType.ESCAPE) {
    override fun compareTo(other: Card): Int = -1
}

object Kraken : Card(CardType.KRAKEN) {
    override fun compareTo(other: Card): Int = -1
}

object WhiteWhale : Card(CardType.WHITE_WHALE) {
    override fun compareTo(other: Card): Int = -1
}

object Butin : Card(CardType.BUTIN) {
    override fun compareTo(other: Card): Int = if (other is Butin) 1 else -1
}

enum class ScaryMaryUsage { ESCAPE, PIRATE, NOT_SET }
data class ScaryMary(val usage: ScaryMaryUsage = ScaryMaryUsage.NOT_SET) : Card(CardType.SCARY_MARY) {
    override fun compareTo(other: Card): Int {
        return when (other) {
            Butin -> 1
            is ColoredCard -> 1
            Escape -> 1
            Kraken -> -1
            is Mermaid -> 1
            is Pirate -> 1
            is ScaryMary -> 1
            SkullkingCard -> -1
            WhiteWhale -> -1
        }
    }

    override fun equals(other: Any?) = other is ScaryMary
    override fun hashCode() = 1
}

data class Deck(val cards: List<Card>, val shuffled: Boolean = true) {
    private val deck = cards
        .let {
            if (shuffled) it.shuffled()
            else it
        }
        .fold(Stack<Card>()) { acc, s -> acc.push(s); acc }

    val size: Int get() = cards.size
    fun pop(): Card = deck.pop()
}
