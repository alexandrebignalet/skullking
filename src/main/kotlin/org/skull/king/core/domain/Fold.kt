package org.skull.king.core.domain

import org.skull.king.core.domain.state.CardPlayed


data class PlayerCard(val playerId: PlayerId, val card: Card)
data class FoldSettlement(
    val nextFoldFirstPlayerId: PlayerId,
    val potentialBonus: Int = 0,
    val won: Boolean = true,
    val butinAllies: List<String> = listOf()
)

enum class BonusType {
    MERMAID_OVER_SKULLKING,
    SKULLKING_PER_PIRATE,
    PIRATE_PER_MERMAID,
    FOURTEEN_BLACK,
    FOURTEEN_DEFAULT,
    BUTIN_WINNER,
}

data class Fold(val cardsByPlayer: Map<PlayerId, Card> = mapOf()) {

    private val values = cardsByPlayer.map { PlayerCard(it.key, it.value) }

    val colorAsked
        get() = values.map(PlayerCard::card).filterIsInstance<ColoredCard>().firstOrNull()?.color

    val size
        get() = cardsByPlayer.size

    fun settle(configuration: GameConfiguration): FoldSettlement {
        val kraken = kraken()
        val whiteWhale = whiteWhale()
        val skullKing = skullKing()
        val pirates = pirates()
        val mermaids = mermaids()
        val highestBlackCard = highestOf(CardColor.BLACK)

        val butinsAllies = butins().map { it.playerId }
        val butinWinnerBonus =
            if (butins().isNotEmpty()) bonusResolver(
                configuration,
                BonusType.BUTIN_WINNER
            ) else 0
        val fourteensBonuses =
            (fourteenBlack()?.let { bonusResolver(configuration, BonusType.FOURTEEN_BLACK) }
                ?: 0) +
                    bonusResolver(
                        configuration,
                        BonusType.FOURTEEN_DEFAULT,
                        fourteensNotBlack().size
                    )

        return when {
            kraken.isNotEmpty() && whiteWhale.isNotEmpty() -> {
                val krakenOrWhiteWhaleFirst =
                    values.first { listOf(CardType.KRAKEN, CardType.WHITE_WHALE).contains(it.card.type) }
                val isWhiteWhaleFirst = krakenOrWhiteWhaleFirst.card.type == CardType.WHITE_WHALE

                return if (isWhiteWhaleFirst) krakenSettlement(configuration, true)
                else whiteWhaleSettlement()
            }

            kraken.isNotEmpty() -> krakenSettlement(configuration)

            whiteWhale.isNotEmpty() -> whiteWhaleSettlement()

            skullKing.isNotEmpty() -> when {
                mermaids.isNotEmpty() -> FoldSettlement(
                    nextFoldFirstPlayerId = mermaids.first().playerId,
                    potentialBonus = bonusResolver(
                        configuration,
                        BonusType.MERMAID_OVER_SKULLKING
                    ) + fourteensBonuses + butinWinnerBonus,
                    butinAllies = butinsAllies
                )

                else -> FoldSettlement(
                    nextFoldFirstPlayerId = skullKing.first().playerId,
                    potentialBonus = bonusResolver(
                        configuration,
                        BonusType.SKULLKING_PER_PIRATE,
                        pirates.size
                    ) + fourteensBonuses + butinWinnerBonus,
                    butinAllies = butinsAllies
                )
            }

            pirates.isNotEmpty() -> FoldSettlement(
                nextFoldFirstPlayerId = pirates.first().playerId,
                potentialBonus =
                bonusResolver(
                    configuration,
                    BonusType.PIRATE_PER_MERMAID,
                    mermaids.size
                ) + fourteensBonuses + butinWinnerBonus,
                butinAllies = butinsAllies
            )

            mermaids.isNotEmpty() -> FoldSettlement(
                mermaids.first().playerId,
                fourteensBonuses + butinWinnerBonus,
                butinAllies = butinsAllies
            )

            highestBlackCard != null -> FoldSettlement(
                highestBlackCard.playerId,
                fourteensBonuses + butinWinnerBonus,
                butinAllies = butinsAllies
            )

            onlyEscapes() -> FoldSettlement(values.first().playerId, 0, false)
            onlyEscapesAndButins() -> {
                val butins = values.filter { it.card.type == CardType.BUTIN }
                FoldSettlement(butins.first().playerId, 0, true)
            }

            else -> FoldSettlement(
                highestOf(colorAsked()!!)!!.playerId,
                fourteensBonuses + butinWinnerBonus,
                butinAllies = butinsAllies
            )
        }
    }

    private fun whiteWhaleSettlement(): FoldSettlement {
        val highestValueColored = highestOfColored()

        return highestValueColored
            ?.let { FoldSettlement(it.playerId, 0, true) }
            ?: FoldSettlement(values.first().playerId, 0, false)
    }

    private fun krakenSettlement(
        configuration: GameConfiguration,
        excludingWhiteWhale: Boolean = false
    ): FoldSettlement {
        val excludedCardsTypes =
            if (excludingWhiteWhale) listOf(CardType.WHITE_WHALE, CardType.KRAKEN) else listOf(CardType.KRAKEN)

        val foldWithoutKraken = values
            .filter { !excludedCardsTypes.contains(it.card.type) }
            .associate { it.playerId to it.card }

        return Fold(foldWithoutKraken)
            .settle(configuration)
            .copy(won = false, potentialBonus = 0)
    }

    private fun kraken() =
        values.filter { it.card.type == CardType.KRAKEN }

    private fun whiteWhale() =
        values.filter { it.card.type == CardType.WHITE_WHALE }

    private fun butins() =
        values.filter { it.card.type == CardType.BUTIN }

    private fun skullKing() =
        values.filter { it.card.type == CardType.SKULLKING }

    private fun mermaids() =
        values.filter { it.card.type == CardType.MERMAID }

    private fun pirates() = values.filter {
        it.card.type == CardType.PIRATE || (it.card is ScaryMary && it.card.usage == ScaryMaryUsage.PIRATE)
    }

    private fun onlyEscapes() = values.all { onlyEscapePredicate(it) }

    private fun onlyEscapePredicate(it: PlayerCard) =
        (it.card.type == CardType.ESCAPE) || (it.card is ScaryMary && it.card.usage == ScaryMaryUsage.ESCAPE)

    private fun onlyEscapesAndButins() = values.all {
        onlyEscapePredicate(it) || it.card.type == CardType.BUTIN
    }

    private fun fourteensNotBlack() =
        values.filter { it.card is ColoredCard && it.card.color != CardColor.BLACK && it.card.value == 14 }

    private fun fourteenBlack() =
        values.find { it.card is ColoredCard && it.card.color == CardColor.BLACK && it.card.value == 14 }

    private fun highestOf(colorAsked: CardColor) =
        values.filter { it.card is ColoredCard && it.card.color == colorAsked }
            .maxByOrNull { (it.card as ColoredCard).value }

    private fun highestOfColored() =
        values.filter { it.card is ColoredCard }.maxByOrNull { (it.card as ColoredCard).value }

    fun colorAsked() =
        values.firstOrNull { it.card is ColoredCard }?.let { (_, card) -> (card as ColoredCard).color }

    private fun bonusResolver(configuration: GameConfiguration, bonusType: BonusType, occurrences: Int = 1) =
        when (configuration) {
            is ClassicConfiguration -> when (bonusType) {
                BonusType.MERMAID_OVER_SKULLKING -> 50
                BonusType.SKULLKING_PER_PIRATE -> 30 * occurrences
                else -> 0
            }

            is BlackRockConfiguration -> when (bonusType) {
                BonusType.MERMAID_OVER_SKULLKING -> 40
                BonusType.SKULLKING_PER_PIRATE -> 30 * occurrences
                BonusType.PIRATE_PER_MERMAID -> 20 * occurrences
                BonusType.FOURTEEN_BLACK -> 20
                BonusType.FOURTEEN_DEFAULT -> 10 * occurrences
                BonusType.BUTIN_WINNER -> 20
            }
        }

    fun receive(event: CardPlayed) = copy(
        cardsByPlayer = cardsByPlayer.plus(Pair(event.playerId, event.card))
    )

    fun cards() = values.map { it.card }
    fun isEmpty() = size == 0
}