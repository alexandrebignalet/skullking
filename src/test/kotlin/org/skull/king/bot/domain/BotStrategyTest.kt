package org.skull.king.bot.domain

import com.fasterxml.jackson.databind.ObjectMapper
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test
import org.skull.king.core.domain.Card
import org.skull.king.core.domain.ClassicConfiguration


class BotStrategyTest {


    @Test
    fun `should calculate each card winning odds against number of player`() {

        val result = (2..6).associateWith { playersCount ->
            val byCard = ClassicConfiguration().deck().cards.associate { card ->
                val odds = winningOddsOf(playersCount, card)

                card.id to "${Math.round(odds * 100)}%"
            }

            byCard
        }

        val jsonResult = ObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(result)
        Approvals.verify(jsonResult)
    }

    fun winningOddsOf(playersCount: Int, card: Card): Double {
        val cards = ClassicConfiguration().deck().cards

        var universe = cards.count()
        var strongerCardsCount = cards.count { it > card }
        var cumulatedOdds = 1.0

        for (i in 1..playersCount) {
            cumulatedOdds *= (universe - strongerCardsCount) / universe.toDouble()
            strongerCardsCount--
            universe--
        }

        return cumulatedOdds
    }

}