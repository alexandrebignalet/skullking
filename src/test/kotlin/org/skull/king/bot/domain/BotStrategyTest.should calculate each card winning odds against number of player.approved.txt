{
  "2" : {
    "COLORED__YELLOW__1" : "1%",
    "COLORED__YELLOW__2" : "1%",
    "COLORED__YELLOW__3" : "1%",
    "COLORED__YELLOW__4" : "2%",
    "COLORED__YELLOW__5" : "2%",
    "COLORED__YELLOW__6" : "3%",
    "COLORED__YELLOW__7" : "3%",
    "COLORED__YELLOW__8" : "4%",
    "COLORED__YELLOW__9" : "5%",
    "COLORED__YELLOW__10" : "5%",
    "COLORED__YELLOW__11" : "6%",
    "COLORED__YELLOW__12" : "7%",
    "COLORED__YELLOW__13" : "8%",
    "COLORED__RED__1" : "1%",
    "COLORED__RED__2" : "1%",
    "COLORED__RED__3" : "1%",
    "COLORED__RED__4" : "2%",
    "COLORED__RED__5" : "2%",
    "COLORED__RED__6" : "3%",
    "COLORED__RED__7" : "3%",
    "COLORED__RED__8" : "4%",
    "COLORED__RED__9" : "5%",
    "COLORED__RED__10" : "5%",
    "COLORED__RED__11" : "6%",
    "COLORED__RED__12" : "7%",
    "COLORED__RED__13" : "8%",
    "COLORED__BLUE__1" : "1%",
    "COLORED__BLUE__2" : "1%",
    "COLORED__BLUE__3" : "1%",
    "COLORED__BLUE__4" : "2%",
    "COLORED__BLUE__5" : "2%",
    "COLORED__BLUE__6" : "3%",
    "COLORED__BLUE__7" : "3%",
    "COLORED__BLUE__8" : "4%",
    "COLORED__BLUE__9" : "5%",
    "COLORED__BLUE__10" : "5%",
    "COLORED__BLUE__11" : "6%",
    "COLORED__BLUE__12" : "7%",
    "COLORED__BLUE__13" : "8%",
    "COLORED__BLACK__1" : "47%",
    "COLORED__BLACK__2" : "49%",
    "COLORED__BLACK__3" : "51%",
    "COLORED__BLACK__4" : "54%",
    "COLORED__BLACK__5" : "56%",
    "COLORED__BLACK__6" : "58%",
    "COLORED__BLACK__7" : "61%",
    "COLORED__BLACK__8" : "63%",
    "COLORED__BLACK__9" : "65%",
    "COLORED__BLACK__10" : "68%",
    "COLORED__BLACK__11" : "71%",
    "COLORED__BLACK__12" : "73%",
    "COLORED__BLACK__13" : "76%",
    "ESCAPE" : "1%",
    "MERMAID__NONE" : "78%",
    "PIRATE__EVIL_EMMY" : "81%",
    "PIRATE__HARRY_THE_GIANT" : "81%",
    "PIRATE__TORTUGA_JACK" : "81%",
    "PIRATE__BADEYE_JOE" : "81%",
    "PIRATE__BETTY_BRAVE" : "81%",
    "SKULLKING" : "93%",
    "SCARY_MARY" : "81%"
  },
  "3" : {
    "COLORED__YELLOW__1" : "0%",
    "COLORED__YELLOW__2" : "0%",
    "COLORED__YELLOW__3" : "0%",
    "COLORED__YELLOW__4" : "0%",
    "COLORED__YELLOW__5" : "0%",
    "COLORED__YELLOW__6" : "0%",
    "COLORED__YELLOW__7" : "1%",
    "COLORED__YELLOW__8" : "1%",
    "COLORED__YELLOW__9" : "1%",
    "COLORED__YELLOW__10" : "1%",
    "COLORED__YELLOW__11" : "1%",
    "COLORED__YELLOW__12" : "2%",
    "COLORED__YELLOW__13" : "2%",
    "COLORED__RED__1" : "0%",
    "COLORED__RED__2" : "0%",
    "COLORED__RED__3" : "0%",
    "COLORED__RED__4" : "0%",
    "COLORED__RED__5" : "0%",
    "COLORED__RED__6" : "0%",
    "COLORED__RED__7" : "1%",
    "COLORED__RED__8" : "1%",
    "COLORED__RED__9" : "1%",
    "COLORED__RED__10" : "1%",
    "COLORED__RED__11" : "1%",
    "COLORED__RED__12" : "2%",
    "COLORED__RED__13" : "2%",
    "COLORED__BLUE__1" : "0%",
    "COLORED__BLUE__2" : "0%",
    "COLORED__BLUE__3" : "0%",
    "COLORED__BLUE__4" : "0%",
    "COLORED__BLUE__5" : "0%",
    "COLORED__BLUE__6" : "0%",
    "COLORED__BLUE__7" : "1%",
    "COLORED__BLUE__8" : "1%",
    "COLORED__BLUE__9" : "1%",
    "COLORED__BLUE__10" : "1%",
    "COLORED__BLUE__11" : "1%",
    "COLORED__BLUE__12" : "2%",
    "COLORED__BLUE__13" : "2%",
    "COLORED__BLACK__1" : "33%",
    "COLORED__BLACK__2" : "35%",
    "COLORED__BLACK__3" : "38%",
    "COLORED__BLACK__4" : "40%",
    "COLORED__BLACK__5" : "43%",
    "COLORED__BLACK__6" : "46%",
    "COLORED__BLACK__7" : "48%",
    "COLORED__BLACK__8" : "51%",
    "COLORED__BLACK__9" : "54%",
    "COLORED__BLACK__10" : "57%",
    "COLORED__BLACK__11" : "61%",
    "COLORED__BLACK__12" : "64%",
    "COLORED__BLACK__13" : "67%",
    "ESCAPE" : "0%",
    "MERMAID__NONE" : "71%",
    "PIRATE__EVIL_EMMY" : "75%",
    "PIRATE__HARRY_THE_GIANT" : "75%",
    "PIRATE__TORTUGA_JACK" : "75%",
    "PIRATE__BADEYE_JOE" : "75%",
    "PIRATE__BETTY_BRAVE" : "75%",
    "SKULLKING" : "91%",
    "SCARY_MARY" : "75%"
  },
  "4" : {
    "COLORED__YELLOW__1" : "0%",
    "COLORED__YELLOW__2" : "0%",
    "COLORED__YELLOW__3" : "0%",
    "COLORED__YELLOW__4" : "0%",
    "COLORED__YELLOW__5" : "0%",
    "COLORED__YELLOW__6" : "0%",
    "COLORED__YELLOW__7" : "0%",
    "COLORED__YELLOW__8" : "0%",
    "COLORED__YELLOW__9" : "0%",
    "COLORED__YELLOW__10" : "0%",
    "COLORED__YELLOW__11" : "0%",
    "COLORED__YELLOW__12" : "0%",
    "COLORED__YELLOW__13" : "1%",
    "COLORED__RED__1" : "0%",
    "COLORED__RED__2" : "0%",
    "COLORED__RED__3" : "0%",
    "COLORED__RED__4" : "0%",
    "COLORED__RED__5" : "0%",
    "COLORED__RED__6" : "0%",
    "COLORED__RED__7" : "0%",
    "COLORED__RED__8" : "0%",
    "COLORED__RED__9" : "0%",
    "COLORED__RED__10" : "0%",
    "COLORED__RED__11" : "0%",
    "COLORED__RED__12" : "0%",
    "COLORED__RED__13" : "1%",
    "COLORED__BLUE__1" : "0%",
    "COLORED__BLUE__2" : "0%",
    "COLORED__BLUE__3" : "0%",
    "COLORED__BLUE__4" : "0%",
    "COLORED__BLUE__5" : "0%",
    "COLORED__BLUE__6" : "0%",
    "COLORED__BLUE__7" : "0%",
    "COLORED__BLUE__8" : "0%",
    "COLORED__BLUE__9" : "0%",
    "COLORED__BLUE__10" : "0%",
    "COLORED__BLUE__11" : "0%",
    "COLORED__BLUE__12" : "0%",
    "COLORED__BLUE__13" : "1%",
    "COLORED__BLACK__1" : "24%",
    "COLORED__BLACK__2" : "26%",
    "COLORED__BLACK__3" : "28%",
    "COLORED__BLACK__4" : "31%",
    "COLORED__BLACK__5" : "33%",
    "COLORED__BLACK__6" : "36%",
    "COLORED__BLACK__7" : "39%",
    "COLORED__BLACK__8" : "42%",
    "COLORED__BLACK__9" : "46%",
    "COLORED__BLACK__10" : "49%",
    "COLORED__BLACK__11" : "53%",
    "COLORED__BLACK__12" : "57%",
    "COLORED__BLACK__13" : "61%",
    "ESCAPE" : "0%",
    "MERMAID__NONE" : "65%",
    "PIRATE__EVIL_EMMY" : "70%",
    "PIRATE__HARRY_THE_GIANT" : "70%",
    "PIRATE__TORTUGA_JACK" : "70%",
    "PIRATE__BADEYE_JOE" : "70%",
    "PIRATE__BETTY_BRAVE" : "70%",
    "SKULLKING" : "91%",
    "SCARY_MARY" : "70%"
  },
  "5" : {
    "COLORED__YELLOW__1" : "0%",
    "COLORED__YELLOW__2" : "0%",
    "COLORED__YELLOW__3" : "0%",
    "COLORED__YELLOW__4" : "0%",
    "COLORED__YELLOW__5" : "0%",
    "COLORED__YELLOW__6" : "0%",
    "COLORED__YELLOW__7" : "0%",
    "COLORED__YELLOW__8" : "0%",
    "COLORED__YELLOW__9" : "0%",
    "COLORED__YELLOW__10" : "0%",
    "COLORED__YELLOW__11" : "0%",
    "COLORED__YELLOW__12" : "0%",
    "COLORED__YELLOW__13" : "0%",
    "COLORED__RED__1" : "0%",
    "COLORED__RED__2" : "0%",
    "COLORED__RED__3" : "0%",
    "COLORED__RED__4" : "0%",
    "COLORED__RED__5" : "0%",
    "COLORED__RED__6" : "0%",
    "COLORED__RED__7" : "0%",
    "COLORED__RED__8" : "0%",
    "COLORED__RED__9" : "0%",
    "COLORED__RED__10" : "0%",
    "COLORED__RED__11" : "0%",
    "COLORED__RED__12" : "0%",
    "COLORED__RED__13" : "0%",
    "COLORED__BLUE__1" : "0%",
    "COLORED__BLUE__2" : "0%",
    "COLORED__BLUE__3" : "0%",
    "COLORED__BLUE__4" : "0%",
    "COLORED__BLUE__5" : "0%",
    "COLORED__BLUE__6" : "0%",
    "COLORED__BLUE__7" : "0%",
    "COLORED__BLUE__8" : "0%",
    "COLORED__BLUE__9" : "0%",
    "COLORED__BLUE__10" : "0%",
    "COLORED__BLUE__11" : "0%",
    "COLORED__BLUE__12" : "0%",
    "COLORED__BLUE__13" : "0%",
    "COLORED__BLACK__1" : "17%",
    "COLORED__BLACK__2" : "19%",
    "COLORED__BLACK__3" : "21%",
    "COLORED__BLACK__4" : "24%",
    "COLORED__BLACK__5" : "26%",
    "COLORED__BLACK__6" : "29%",
    "COLORED__BLACK__7" : "32%",
    "COLORED__BLACK__8" : "35%",
    "COLORED__BLACK__9" : "39%",
    "COLORED__BLACK__10" : "43%",
    "COLORED__BLACK__11" : "47%",
    "COLORED__BLACK__12" : "51%",
    "COLORED__BLACK__13" : "56%",
    "ESCAPE" : "0%",
    "MERMAID__NONE" : "61%",
    "PIRATE__EVIL_EMMY" : "67%",
    "PIRATE__HARRY_THE_GIANT" : "67%",
    "PIRATE__TORTUGA_JACK" : "67%",
    "PIRATE__BADEYE_JOE" : "67%",
    "PIRATE__BETTY_BRAVE" : "67%",
    "SKULLKING" : "93%",
    "SCARY_MARY" : "67%"
  },
  "6" : {
    "COLORED__YELLOW__1" : "0%",
    "COLORED__YELLOW__2" : "0%",
    "COLORED__YELLOW__3" : "0%",
    "COLORED__YELLOW__4" : "0%",
    "COLORED__YELLOW__5" : "0%",
    "COLORED__YELLOW__6" : "0%",
    "COLORED__YELLOW__7" : "0%",
    "COLORED__YELLOW__8" : "0%",
    "COLORED__YELLOW__9" : "0%",
    "COLORED__YELLOW__10" : "0%",
    "COLORED__YELLOW__11" : "0%",
    "COLORED__YELLOW__12" : "0%",
    "COLORED__YELLOW__13" : "0%",
    "COLORED__RED__1" : "0%",
    "COLORED__RED__2" : "0%",
    "COLORED__RED__3" : "0%",
    "COLORED__RED__4" : "0%",
    "COLORED__RED__5" : "0%",
    "COLORED__RED__6" : "0%",
    "COLORED__RED__7" : "0%",
    "COLORED__RED__8" : "0%",
    "COLORED__RED__9" : "0%",
    "COLORED__RED__10" : "0%",
    "COLORED__RED__11" : "0%",
    "COLORED__RED__12" : "0%",
    "COLORED__RED__13" : "0%",
    "COLORED__BLUE__1" : "0%",
    "COLORED__BLUE__2" : "0%",
    "COLORED__BLUE__3" : "0%",
    "COLORED__BLUE__4" : "0%",
    "COLORED__BLUE__5" : "0%",
    "COLORED__BLUE__6" : "0%",
    "COLORED__BLUE__7" : "0%",
    "COLORED__BLUE__8" : "0%",
    "COLORED__BLUE__9" : "0%",
    "COLORED__BLUE__10" : "0%",
    "COLORED__BLUE__11" : "0%",
    "COLORED__BLUE__12" : "0%",
    "COLORED__BLUE__13" : "0%",
    "COLORED__BLACK__1" : "13%",
    "COLORED__BLACK__2" : "14%",
    "COLORED__BLACK__3" : "16%",
    "COLORED__BLACK__4" : "19%",
    "COLORED__BLACK__5" : "21%",
    "COLORED__BLACK__6" : "24%",
    "COLORED__BLACK__7" : "27%",
    "COLORED__BLACK__8" : "30%",
    "COLORED__BLACK__9" : "34%",
    "COLORED__BLACK__10" : "38%",
    "COLORED__BLACK__11" : "42%",
    "COLORED__BLACK__12" : "47%",
    "COLORED__BLACK__13" : "52%",
    "ESCAPE" : "0%",
    "MERMAID__NONE" : "58%",
    "PIRATE__EVIL_EMMY" : "64%",
    "PIRATE__HARRY_THE_GIANT" : "64%",
    "PIRATE__TORTUGA_JACK" : "64%",
    "PIRATE__BADEYE_JOE" : "64%",
    "PIRATE__BETTY_BRAVE" : "64%",
    "SKULLKING" : "96%",
    "SCARY_MARY" : "64%"
  }
}